zrml/neo-swaps/src/types/liquidity_tree.rs

```diff
@@ -0,0 +1,1205 @@
+        );
+    }
+
+    #[test]
+    fn exit_fails_on_unclaimed_fees_at_root() {
+        let mut tree = utility::create_test_tree();
+        // Clear unclaimed fees except for root.
+        tree.nodes[0].lazy_fees = _1;
+        tree.nodes[1].lazy_fees = Zero::zero();
+        tree.nodes[3].fees = Zero::zero();
+        tree.nodes[3].lazy_fees = Zero::zero();
+        assert_err!(tree.exit(&5, 1), LiquidityTreeError::UnclaimedFees.into_dispatch::<Runtime>());
+    }
+
+    #[test]
+    fn exit_fails_on_unclaimed_fees_on_middle_of_path() {
+        let mut tree = utility::create_test_tree();
+        // Clear unclaimed fees except for the middle node.
+        tree.nodes[3].fees = Zero::zero();
+        tree.nodes[3].lazy_fees = Zero::zero();
+        assert_err!(tree.exit(&5, 1), LiquidityTreeError::UnclaimedFees.into_dispatch::<Runtime>());
+    }
+
+    #[test]
+    fn exit_fails_on_unclaimed_fees_at_last_node_due_to_lazy_fees() {
+        let mut tree = utility::create_test_tree();
+        // Clear unclaimed fees except for the last node.
+        tree.nodes[1].lazy_fees = Zero::zero();
+        // This ensures that the error is caused by propagated lazy fees sitting in the node.
+        tree.nodes[3].fees = Zero::zero();
+        assert_err!(tree.exit(&5, 1), LiquidityTreeError::UnclaimedFees.into_dispatch::<Runtime>());
+    }
+
+    #[test]
+    fn exit_fails_on_unclaimed_fees_at_last_node_due_to_fees() {
+        let mut tree = utility::create_test_tree();
+        // Clear unclaimed fees except for the last node.
+        tree.nodes[1].lazy_fees = Zero::zero();
+        // This ensures that the error is caused by normal fees.
+        tree.nodes[3].lazy_fees = Zero::zero();
+        assert_err!(tree.exit(&5, 1), LiquidityTreeError::UnclaimedFees.into_dispatch::<Runtime>());
+    }
+
+    #[test]
+    fn deposit_fees_works_root() {
+        let mut tree = utility::create_test_tree();
+        let mut nodes = tree.nodes.clone().into_inner();
+        let account_to_index = tree.account_to_index.clone().into_inner();
+        let abandoned_nodes = tree.abandoned_nodes.clone().into_inner();
+        let amount = _12;
+        nodes[0].lazy_fees += amount;
+        tree.deposit_fees(amount).unwrap();
+        assert_liquidity_tree_state!(tree, nodes, account_to_index, abandoned_nodes);
+    }
+
+    #[test]
+    fn deposit_fees_works_no_root() {
+        let mut tree = utility::create_test_tree();
+        tree.nodes[0].account = None;
+        tree.nodes[1].stake = Zero::zero();
+        tree.nodes[2].fees = Zero::zero();
+        let mut nodes = tree.nodes.clone().into_inner();
+        let account_to_index = tree.account_to_index.clone().into_inner();
+        let abandoned_nodes = tree.abandoned_nodes.clone().into_inner();
+        let amount = _12;
+        nodes[0].lazy_fees += amount;
+        tree.deposit_fees(amount).unwrap();
+        assert_liquidity_tree_state!(tree, nodes, account_to_index, abandoned_nodes);
+    }
```

Comment:
What should this test exactly show? It seems like the same as `deposit_fees_works_root`.

Reply:
The difference is that in `no_root` the root is not owned by an LP. I thought it was necessary to ensure that this doesn't hamper fee distribution.